informer/informer.go

```diff
@@ -0,0 +1,383 @@
+	// ResyncPeriod is the time to wait before releasing update events again.
+	ResyncPeriod time.Duration
+}
+
+// DefaultConfig provides a default configuration to create a new by best
+// effort.
+func DefaultConfig() Config {
+	return Config{
+		// Dependencies.
+		BackOff:    nil,
+		RestClient: nil,
+
+		// Settings.
+		RateWait:     DefaultRateWait,
+		ResyncPeriod: DefaultResyncPeriod,
+	}
+}
+
+// Informer implements primitives to watch event objects from the Kubernetes API
+// in a deterministic way.
+type Informer struct {
+	// Dependencies.
+	backOff    backoff.BackOff
+	restClient rest.Interface
+
+	// Internals.
+	cache       *sync.Map
+	initializer chan struct{}
+
+	// Settings.
+	rateWait     time.Duration
+	resyncPeriod time.Duration
+}
+
+// New creates a new Informer.
+func New(config Config) (*Informer, error) {
+	// Dependencies.
+	if config.BackOff == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.BackOff must not be empty")
+	}
+	if config.RestClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.RestClient must not be empty")
+	}
+
+	// Settings.
+	if config.ResyncPeriod == 0 {
+		return nil, microerror.Maskf(invalidConfigError, "config.ResyncPeriod must not be empty")
+	}
+
+	newInformer := &Informer{
+		// Settings.
+		backOff:    config.BackOff,
+		restClient: config.RestClient,
+
+		// Internals.
+		cache:       &sync.Map{},
+		initializer: make(chan struct{}),
+
+		// Settings.
+		rateWait:     config.RateWait,
+		resyncPeriod: config.ResyncPeriod,
+	}
+
+	return newInformer, nil
+}
+
+// Watch only watches objects using a stream decoder. Afer the resync period the
+// active watch is closed and a new stream decoder watches the API again. This
+// mechanism has a very small potential to not recognize delete events of
+// objects that do not use finalizers.
+//
+// Watch takes a context as first argument which can be used to cancel the
+// watch. The second argument provided is the raw API endpoint path the watcher
+// is using to fetch event objects. The third argument is a custom object
+// factory to create zero objects on demand. The watcher is using this to decode
+// event objects.
+//
+// Watch returns channels for delete, update and error events, in this order.
+// Events will be dispatched as soon as they happen.
+//
+// That the resync period configured for the informer will trigger periodic
+// updates of event objects via the update channel.
+func (i *Informer) Watch(ctx context.Context, endpoint string, factory ZeroObjectFactory) (chan watch.Event, chan watch.Event, chan error) {
+	done := make(chan struct{}, 1)
+	eventChan := make(chan watch.Event, 1)
+
+	deleteChan := make(chan watch.Event, 1)
+	updateChan := make(chan watch.Event, 1)
+	errChan := make(chan error, 1)
+
+	go func() {
+		for {
+			select {
+			case <-done:
+			case event := <-eventChan:
+				switch event.Type {
+				case watch.Added:
+					err := i.cacheAndReleaseIfNotExists(event, updateChan)
```

Comment:
I would like to change this to `cacheAndSendIfNotExists`

Reply:
Done.